source/lac/scalapack.cc

```diff
@@ -0,0 +1,944 @@
+
+  // processor grid order.
+  // FIXME: default to column major?
+  const bool column_major = false;
+
+  // Initialize Cblas context from the provided communicator
+  blacs_context = Csys2blacs_handle(mpi_communicator);
+  const char *order = ( column_major ? "Col" : "Row" );
+  // FIXME: blacs_context can be modified below. Thus Cblacs2sys_handle
+  // may not return the same MPI communicator
+  Cblacs_gridinit(&blacs_context, order, n_process_rows, n_process_columns);
+
+  // Blacs may modify the grid size on processes which are not used
+  // in the grid. So provide copies below:
+  int procrows_ = n_process_rows;
+  int proccols_ = n_process_columns;
+  Cblacs_gridinfo( blacs_context, &procrows_, &proccols_, &this_process_row, &this_process_column );
+
+  // If this MPI core is not on the grid, flag is as inactive and
+  // skip all jobs
+  // FIXME: different condition is used here
+  // https://stackoverflow.com/questions/18516915/calling-blacs-with-more-processes-than-used
+  if (this_process_row < 0 || this_process_column < 0)
+    active = false;
+  else
+    active = true;
+
+  // Create an auxiliary communicator which has root and all inactive cores
+  // Assume that inactive cores start with id=n_process_rows*n_process_columns
+  Assert (active || this_mpi_process >= n_process_rows*n_process_columns,
+          ExcInternalError());
+
+  std::vector<int> inactive_with_root_ranks;
+  inactive_with_root_ranks.push_back(0);
+  for (int i = n_process_rows*n_process_columns; i < n_mpi_processes; ++i)
+    inactive_with_root_ranks.push_back(i);
+
+  // Get the group of processes in mpi_communicator
+  int ierr = 0;
+  MPI_Group all_group;
+  ierr = MPI_Comm_group(mpi_communicator, &all_group);
+  AssertThrowMPI(ierr);
+
+  // Construct the group containing all ranks we need:
+  MPI_Group inactive_with_root_group;
+  const int n = inactive_with_root_ranks.size();
+  ierr = MPI_Group_incl(all_group,
+                        n, &inactive_with_root_ranks[0],
+                        &inactive_with_root_group);
+  AssertThrowMPI(ierr);
+
+  // Create the communicator based on the group
+  // Note that, on most cores the communicator will be MPI_COMM_NULL
+  // FIXME: switch to MPI_Comm_create_group for MPI-3 so that only processes within the to-be subgroup call this
+  ierr = MPI_Comm_create(mpi_communicator, inactive_with_root_group,
+                         &mpi_communicator_inactive_with_root);
+  AssertThrowMPI(ierr);
+
+  MPI_Group_free(&all_group);
+  MPI_Group_free(&inactive_with_root_group);
+
+  // Double check that the process with rank 0 in subgroup is active:
+#ifdef DEBUG
+  if (mpi_communicator_inactive_with_root != MPI_COMM_NULL)
+    {
+      int subgroup_rank = -1, subgroup_size = -1;
+      MPI_Comm_rank(mpi_communicator_inactive_with_root, &subgroup_rank);
+      MPI_Comm_size(mpi_communicator_inactive_with_root, &subgroup_size);
```

Comment:
Use `Utilities::MPI::this_mpi_process(...)` and `Utilities::MPI::n_mpi_processes(...)`, and then you can inline these calls into the assertion.